### PR TITLE
improve: replace server config validation

### PR DESCRIPTION
**Pull request by dnephin**
Branched from #2611

With internal/validate

Also required adding slice traversal to internal/validate

## Related Issues

Related to #2583

### CODE REVIEW

File: internal/server/config.go [hunk 1]
```diff
@@ -19,55 +18,106 @@ import (
...
+	ClientID     string
+	ClientSecret string
 	Kind         string
 	AuthURL      string
 	Scopes       []string
 }
 
+func (p Provider) ValidationRules() []validate.ValidationRule {
+	return []validate.ValidationRule{
+		// TODO: api.ValidateName
```
Comment (jmorganca) [inline internal/server/config.go:37]: Excited for this: consistent naming validation is so important for APIs

**Review by jmorganca — APPROVED** [review]

File: internal/server/config.go (hunk 1 above)
Comment (dnephin) [reply internal/server/config.go]: I pushed another commit which resolves this TODO

**Review by dnephin — COMMENTED** [review]
Added some local test coverage in the `internal/validate` package for slice traversal. 

Also changed the default field name (when there are no tags) to use the field name with the first letter lowercased.

File: internal/server/config.go [hunk 2]
```diff
@@ -9,65 +9,116 @@ import (
...
+}
+
 type Grant struct {
-	User     string `validate:"excluded_with=Group,excluded_with=Machine"`
-	Group    string `validate:"excluded_with=User,excluded_with=Machine"`
-	Machine  string `validate:"excluded_with=User,excluded_with=Group"` // deprecated
-	Resource string `validate:"required"`
+	User     string
+	Group    string
+	Machine  string // deprecated
```
Comment (mxyng) [inline internal/server/config.go:49]: Can probably remove `machine` now

**Review by mxyng — APPROVED** [review]

File: internal/server/config.go (hunk 2 above)
Comment (dnephin) [reply internal/server/config.go:49]: I was wondering about that. Going to leave it for a follow up in case there are other implications to removing it.

**Pull request merged**